jerry-debugger/src/lib/debugger-client.ts

```diff
@@ -0,0 +1,404 @@
+      this.abortConnection('message too short');
+    }
+
+    if (this.cpointerSize === 0) {
+      if (message[0] !== S.JERRY_DEBUGGER_CONFIGURATION || message.byteLength !== 5) {
+        this.abortConnection('the first message must be configuration');
+      }
+
+      this.maxMessageSize = message[1];
+      this.cpointerSize = message[2];
+      this.littleEndian = Boolean(message[3]);
+
+      this.version = message[4];
+
+      if (this.cpointerSize !== 2 && this.cpointerSize !== 4) {
+        this.abortConnection('compressed pointer must be 2 or 4 bytes long');
+      }
+
+      if (this.version !== JERRY_DEBUGGER_VERSION) {
+        this.abortConnection('incorrect target debugger version detected: ' + this.version
+                             + ' expected: ' + JERRY_DEBUGGER_VERSION);
+      }
+
+      return;
+    }
+
+    switch (message[0]) {
+      case S.JERRY_DEBUGGER_PARSE_ERROR:
+        console.log('[Parse Error]');
+        console.log('debugger parse error');
+        this.parseObj = undefined;
+        return;
+
+      case S.JERRY_DEBUGGER_BYTE_CODE_CP:
+      case S.JERRY_DEBUGGER_PARSE_FUNCTION:
+      case S.JERRY_DEBUGGER_BREAKPOINT_LIST:
+      case S.JERRY_DEBUGGER_BREAKPOINT_OFFSET_LIST:
+      case S.JERRY_DEBUGGER_SOURCE_CODE:
+      case S.JERRY_DEBUGGER_SOURCE_CODE_END:
+      case S.JERRY_DEBUGGER_SOURCE_CODE_NAME:
+      case S.JERRY_DEBUGGER_SOURCE_CODE_NAME_END:
+      case S.JERRY_DEBUGGER_FUNCTION_NAME:
+      case S.JERRY_DEBUGGER_FUNCTION_NAME_END:
+        if (!this.parseObj) {
+          this.parseObj = new ParsedSource(this);
+        }
+        this.parseObj.receive(message);
+        return;
+
+      case S.JERRY_DEBUGGER_RELEASE_BYTE_CODE_CP:
+        console.log('[Byte Code CP]');
+        this.releaseFunction(message);
+        return;
+
+      case S.JERRY_DEBUGGER_MEMSTATS_RECEIVE: {
+        console.log('[Mem Stats]');
+        const messagedata = this.decodeMessage('IIIII', message, 1);
+
+        console.log('Allocated bytes: ' + messagedata[0]);
+        console.log('Byte code bytes: ' + messagedata[1]);
+        console.log('String bytes: ' + messagedata[2]);
+        console.log('Object bytes: ' + messagedata[3]);
+        console.log('Property bytes: ' + messagedata[4]);
+        return;
+      }
+
+      case S.JERRY_DEBUGGER_BREAKPOINT_HIT:
+      case S.JERRY_DEBUGGER_EXCEPTION_HIT: {
+        console.log('[Breakpoint Hit]');
+        var breakpointData = this.decodeMessage('CI', message, 1);
+        var breakpointRef = this.getBreakpoint(breakpointData);
+        var breakpoint = breakpointRef.breakpoint;
+
+        if (message[0] == S.JERRY_DEBUGGER_EXCEPTION_HIT) {
+          console.log('Exception throw detected');
+          if (this.exceptionData) {
+            console.log('Exception hint: ' + cesu8ToString(this.exceptionData));
+            this.exceptionData = undefined;
+          }
+        }
+
+        this.lastBreakpointHit = breakpoint;
```

Comment:
Could you cleanly separate the message parsing vs debugger client state tracking?

I would move the parsing into a different file. Also easier to unit test that way.

Reply:
OK this turned out to just need at least one method in the interface to be actually defined.

Reply:
OK, glad you figured it out.